### ISSUE TITLE
Moving vars/function to a different namespace

### ISSUE THREAD

**Issue by martinklepsch** [clojure-lsp/clojure-lsp#566]
After a coding spree I often realize that I'd like to rearrange some of my code. For instance I defined a utility function and would now like to also move that to the "utilities" namespace. 

In these situations it would be create to trigger a command and basically rename a var including it's namespace portion.  For example I'd like to move `my.app.core/find-first` to `my.app.util/find-first`. Besides just moving the var it would also be great if all call sites are updated accordingly and requires to `my.app.util` are added as needed.

I can't think of many alternatives other than just moving the code manually, which is tedious and error-prone. That said there probably are lots of IDEs that support this type of functionality and it might be good to take a look their UX for inspiration.

**Comment by ericdallo** [comment]
Thank you for the suggestion, it makes sense indeed, I think we can rely on the existing `rename` feature, where the user can just append the new namespace and clojure-lsp would take care of the rest

**Comment by mynomoto** [comment]
> Thank you for the suggestion, it makes sense indeed, I think we can rely on the existing rename feature, where the user can just append the new namespace and clojure-lsp would take care of the rest

Does this mean that I would find a usage of the var and rename it? Some questions:
1. Will it understand aliases? How would be the behavior?
2. Would it change all usages in the project?

If 2 is true it should be enough to be helpful however 1 works. I think something to make aliases consistent could be missing but that should be other issue anyway.

**Comment by ericdallo** [comment]
yeah, maybe rename wouldn't work for alias cases but for 2 yes, maybe we can start with a code action that would be available when hovering over function definition?

**Comment by mynomoto** [comment]
One example (not trivial and only a little tricky):
```clojure
; src/app/a.clj
(ns app.a)

(defn foo
  []
  :foo)

(defn another
  []
  :another)

; src/app/b.clj
(ns app.b
  (:require
    [app.a :as a]))

(defn bar
  []
  (a/foo))


; src/app/c.clj
(ns app.c
  (:require
    [app.a :as a.a]))

(defn cde
  []
  (a.a/foo))


; src/app/d.clj
(ns app.d
  (:require
    [app.a :as a.a]
    [app.b :as b]))

(defn fgh
  []
  (= (a.a/foo) (b/bar)))
```

The user would hover over `foo` on `app.a` and select "move to other namespace" (improvising here, details may vary). It would open an input for the user to choose the namespace. They will pick `app.b`. What I would expect to happen:
```clojure
; src/app/a.clj
(ns app.a)
; foo was removed
(defn another
  []
  :another)


; src/app/b.clj
(ns app.b
  (:require
    [app.a :as a]))

(defn foo ; here is foo
  []
  :foo)

(defn bar
  []
  (foo)) ; this call was adjusted


; src/app/c.clj
(ns app.c
  (:require
    [app.a :as a.a]
    [app.b :as app.b])) ; this is a new require

(defn cde
  []
  (app.b/foo)) ; this call was adjusted


; src/app/d.clj
(ns app.d
  (:require
    [app.a :as a.a]
    [app.b :as b]))

(defn fgh
  []
  (= (b/foo) (b/bar))) ; the foo call was adjusted
```

Writing the example was not as trivial as I would expect, no idea if this is feasible. Ideally unused namespaces could be removed but some uses have namespaces with side effects so removing would not be always correct.

Is this something like this that you have in your mind @ericdallo ?

**Comment by ericdallo** [comment]
Yes, this example brings the complexity of this feature which is having a lot of corner cases like:
- we should move the function to the other namespace but we can't add it always on the bottom of the ns since there could be usages above so we could need to check all previous usages and make sure we move it to above the first usage in the ns or to the bottom if none
- we would need to add the require if there is none yet, if no required yet, what alias should we use if the ns was never required before? maybe use the last segment of the namespace
- change all calls on `app.b` to remove the alias since the function is on the same ns (one more corner case I didn't realize until now)

It's certainly possible, the issue is the need to handle all those cases, we could start with some simpler support and improve eventually, although making a partial feature makes user don't rely on that unfortunately

**Comment by mynomoto** [comment]
A few more complications I got while coming up with that example but added to much complexity for what should be a simple example.
- the moved function could use functions from the destination namespace. If that was not the case you could always add it on the top of the namespace.
- the origin namespace could become empty
- you need to add requires to namespaces used by the moved function on the destination namespace. The destination namespace could have requires using different aliases.
- moving the function could cause a circular dependency.

**Comment by snoe** [comment]
> although making a partial feature makes user don't rely on that unfortunately

It's actually very important to fully move, including aliases and everything. It's not just that they can't rely on it, but people will use it, without realizing the corner cases and break their project in hard to find ways (lots of changes across the project and things not changed will be hard to spot).

**Comment by orestis** [comment]
Chiming in: Perhaps it'd be worthwhile to implement "checks" that will bail out of the action in the complex cases. 

So e.g. you could implement a simple scenario of moving a function to e.g. a util namespace because you want to use it in other places, that will just put it at the bottom and add a require and change the references/usages. If there's any complex cases you could throw an error saying "can't handle this yet". 

It's a feature that could get progressively better but probably is extremely hard to get 100% right from the get go.

**Comment by ericdallo** [comment]
Yes, one of the most chalenging is to require and update the references, there are some issues mentioned above, and this is required to the most simple usage of this feature

**Comment by snoe** [comment]
This is called `move-form` https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-move-form

**Comment by orestis** [comment]
I just went through a round of this yesterday. What I end up doing (manually) is actually start from the actual function I want to move (without considering its own dependencies):

1. find or create a new namespace
2. move the function from old to new namespace
3. Rename references that point to the moved function (in the old namespace and perhaps others that were referencing), potentially adding requires as needed. 

Then I go to consider the actual moved function. If clj-kondo complains of unresolved vars, I will pick one at random and repeat steps 1-3. Repeat until everything is fine. I had to make a few functions public too. 

Steps 1-3 are tedious to do and where LSP could make it easier. I’m happy looking at the innards of a moved function myself since usually there’s judgment calls to make there. 

I also don’t mind circular dependencies since I will catch them manually at the thinking phase, but LSP or Kondo warning about them statically would be a major help.

**Comment by gilch** [comment]
I tried out the `lsp-move-form`. It's not working right yet. I discussed this a bit with @ericdallo on Clojurians Slack, and it sounds like finding the corner cases will help improve this. Issues I've noticed so far:

- doesn't add references the moved form needs to the destination ns.
  - I think this one is already known. It's pretty obvious.
  - clj-refactor.el (a CIDER addition) already does this part properly. Maybe reading that code would help.
- deletes aliases that were still in use when a var using the same alias needs a rename.
  - To be clear, this was in neither the source nor the destination ns.
- somehow got a blank alias in the source ns, causing syntax errors.
- attempting to move a private form failed silently
  - should probably cycle privacy first, or at least generate a sensible error

One possible approach I thought of would be to start with the fully-qualified names for everything that needs renaming. (External references to the moved var, and vars referenced inside the move form.) Everything would work without worrying about aliases and refers, as long as the necessary require is present. Clojure's syntax quote takes advantage of this feature to make macroexpanions work across namespaces. Smarter renames could then be added later as we think of them, but anything we miss will still work with a fully-qualified name.

`lsp-move-form` also need not delete any existing requires, as we have other ways to clean those up that check to see if they're unused first.

**Comment by ericdallo** [comment]
Thanks for the deep dive @gilch, it seems to me that we could focus on fixing each one of these issues separately, right?
WDYT about we closing this issue and creating a new one for each known issue as you listed with code examples? I think this could help us understand and discuss which one are easy to implement and which ones we should discuss more about how to do.

> One possible approach I thought of would be to start with the fully-qualified names for everything that needs renaming. (External references to the moved var, and vars referenced inside the move form.) Everything would work without worrying about aliases and refers, as long as the necessary require is present. Clojure's syntax quote takes advantage of this feature to make macroexpanions work across namespaces. Smarter renames could then be added later as we think of them, but anything we miss will still work with a fully-qualified name.

It's an alternative indeed, but clojure-lsp already has smart features to add-missing-require, being an alias or refer, so we could leverage that and fallback to your alternative if no alias/refer is found.

**Comment by gilch** [comment]
I also noticed a case where `lsp-clojure-move-form` added a require with a metadata map, which seemed weird. I haven't been able to reproduce it, but I might have forgotten to save files when it happened.
